### PR TITLE
Bugs/stagedamage

### PR DESCRIPTION
**Pull request by rnugent3**

### CODE REVIEW

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 1]
```diff
@@ -56,7 +56,7 @@ public class ImpactAreaStageDamage : Validation, IReportMessage
             _GraphicalFrequency = graphicalFrequency;
             _DischargeStage = dischargeStage;
             _ImpactAreaID = impactAreaID;
-            _inventory = inventory.GetInventoryTrimmmedToPolygon(impactAreaID);
+            //_inventory = inventory.GetInventoryTrimmmedToPolygon(impactAreaID);
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs]: this is no longer commented out

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 2]
```diff
@@ -280,12 +280,14 @@ public List<UncertainPairedData> Compute(IProvideRandomNumbers randomProvider)
...
         }
-        public float[] ExtrapolateFromAboveAtIndexLocation(float[] stagesAtStructuresHighestProfile, float upperInterval, int stepCount)
+        public static float[] ExtrapolateFromAboveAtIndexLocation(float[] stagesAtStructuresHighestProfile, float upperInterval, int stepCount)
         {
             float[] extrapolatedStages = new float[stagesAtStructuresHighestProfile.Length];
+            int i = 0;
             foreach (float structureStage in stagesAtStructuresHighestProfile)
             {
-                extrapolatedStages[stepCount] = structureStage + upperInterval * stepCount;
+                extrapolatedStages[i] = structureStage + upperInterval * stepCount;
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:290]: bug fix

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 3]
```diff
@@ -280,12 +280,14 @@ public List<UncertainPairedData> Compute(IProvideRandomNumbers randomProvider)
             List<UncertainPairedData> results = ConsequenceDistributionResults.ToUncertainPairedData(allStagesAtIndexLocation, consequenceDistributionResults);
             return results;
         }
-        public float[] ExtrapolateFromAboveAtIndexLocation(float[] stagesAtStructuresHighestProfile, float upperInterval, int stepCount)
+        public static float[] ExtrapolateFromAboveAtIndexLocation(float[] stagesAtStructuresHighestProfile, float upperInterval, int stepCount)
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:284]: I don't think this is static anymore

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 4]
```diff
@@ -37,8 +37,8 @@ public class ImpactAreaStageDamage : Validation, IReportMessage
         private ConvergenceCriteria convergenceCriteria;
 
         private int seed = 1234;
-        private int _numInterpolatedStagesToCompute = 10;
-        private int _numExtrapolatedStagesToCompute = 50;
+        private int _numInterpolatedStagesToCompute = 1;
+        private int _numExtrapolatedStagesToCompute = 5;
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:40]: I think we should bump these back up later

File: fda-model/stageDamage/ImpactAreaStageDamage.cs (hunk 3 above)
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:284]: static for testing

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 5]
```diff
@@ -291,12 +291,14 @@ public static float[] ExtrapolateFromAboveAtIndexLocation(float[] stagesAtStruct
             }
             return extrapolatedStages;
         }
-        public float[] ExtrapolateFromBelowStagesAtIndexLocation(float[] WSEsAtLowest, float interval, int i)
+        public static float[] ExtrapolateFromBelowStagesAtIndexLocation(float[] WSEsAtLowest, float interval, int i, int numInterpolatedStagesToCompute)
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:295]: because we went with static

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 6]
```diff
@@ -291,12 +291,14 @@ public static float[] ExtrapolateFromAboveAtIndexLocation(float[] stagesAtStruct
...
         {
             float[] extrapolatedStages = new float[WSEsAtLowest.Length];
+            int j = 0;
             foreach (float stage in WSEsAtLowest)
             {
-                extrapolatedStages[i] = stage - interval * (_numInterpolatedStagesToCompute - i);
+                extrapolatedStages[j] = stage - interval * (numInterpolatedStagesToCompute - i);
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:301]: bug fix

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 7]
```diff
@@ -308,7 +310,7 @@ public static ConsequenceDistributionResults ComputeDamageOneCoordinate(IProvide
             double upperProb = .975;
             ConsequenceDistributionResults consequenceDistributionResults = new ConsequenceDistributionResults(convergenceCriteria);
             long iteration = 0;
-            while (consequenceDistributionResults.ResultsAreConverged(upperProb, lowerProb))
+            while (!consequenceDistributionResults.ResultsAreConverged(upperProb, lowerProb))
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs]: bug fix. the convergence test is broken out later.

File: fda-model/metrics/ConsequenceDistributionResults.cs [hunk 8]
```diff
@@ -129,7 +145,15 @@ private void AddConsequenceRealizationByAssetCategory(double consequenceRealizat
             ConsequenceDistributionResult structureConsequenceDistributionResult = GetConsequenceResult(damageCategory, assetCategory, impactAreaID);
             if (structureConsequenceDistributionResult.IsNull)
             {
-                ConsequenceDistributionResult newStructureDamageResult = new ConsequenceDistributionResult(damageCategory, assetCategory, _ConvergenceCriteria, impactAreaID);
+                //HACK: hacking in bin width 
+                //TODO: Get rid of this terrible coding
+                double binWidth = 500;
```
Comment (rnugent3) [inline fda-model/metrics/ConsequenceDistributionResults.cs:150]: this needs to be fixed with priority

File: fda-model/metrics/ConsequenceResults.cs [hunk 9]
```diff
@@ -36,8 +36,6 @@ public bool IsNull
         public ConsequenceResults()
         {
             _consequenceResultList = new List<ConsequenceResult>();
-            ConsequenceResult dummyConsequenceResult = new ConsequenceResult();
```
Comment (rnugent3) [inline fda-model/metrics/ConsequenceResults.cs:39]: this was causing the "unassigned" functions

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 10]
```diff
@@ -281,46 +281,73 @@ public List<UncertainPairedData> Compute(IProvideRandomNumbers randomProvider)
...
             double upperProb = .975;
-            ConsequenceDistributionResults consequenceDistributionResults = new ConsequenceDistributionResults(convergenceCriteria);
-            long iteration = 0;
-            while (consequenceDistributionResults.ResultsAreConverged(upperProb, lowerProb))
+            
+            bool resultsAreNotConverged = true;
+            long expectedIterations = convergenceCriteria.MaxIterations;
+            long iterations = convergenceCriteria.MinIterations;
+            List<ConsequenceResults> results = new List<ConsequenceResults>();
+            while (resultsAreNotConverged)
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:318]: we will want to parallelize this

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 11]
```diff
@@ -281,46 +281,73 @@ public List<UncertainPairedData> Compute(IProvideRandomNumbers randomProvider)
...
             }
-            return consequenceDistributionResults;
+            return returnValue;
         }
         public void ReportMessage(object sender, MessageEventArgs e)
         {
             MessageReport?.Invoke(sender, e);
         }
-
+        public void ReportProgress(object sender, ProgressReportEventArgs e)
```
Comment (rnugent3) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:347]: this needs to be used

**Comment by Brennan1994** [comment]
@rnugent3 I'm not tagged to review this, is this something you're wanting to merge back in and got forgotten or is it still a work in progress?

**Comment by rnugent3** [comment]
@Brennan1994 I thought I tagged you! Oops. Please review when you have a chance.

File: fda-model/metrics/ConsequenceDistributionResults.cs (hunk 8 above)
Comment (Brennan1994) [reply fda-model/metrics/ConsequenceDistributionResults.cs:150]: This does seem distinctly not ideal. Do you have what you need to fix this before we merge?

File: fda-model/metrics/ConsequenceDistributionResults.cs [hunk 12]
```diff
@@ -286,16 +310,17 @@ public double ConsequenceExceededWithProbabilityQ(double exceedanceProbability,
         /// <returns></returns>
         public ConsequenceDistributionResult GetConsequenceResult(string damageCategory, string assetCategory, int impactAreaID)
         {
-            foreach (ConsequenceDistributionResult damageResult in _consequenceResultList)
+            //foreach (ConsequenceDistributionResult damageResult in _consequenceResultList)
+            for (int i = 0; i < _consequenceResultList.Count; i++)
             {
                 //The impact area should always be equal because a consequence result reflects 1 impact area and a consequence resultS reflects 1 impact area   
-                if (damageResult.RegionID.Equals(impactAreaID))
+                if (_consequenceResultList[i].RegionID.Equals(impactAreaID))
```
Comment (Brennan1994) [inline fda-model/metrics/ConsequenceDistributionResults.cs:317]: In general, if we've validated something upstream (like only using one impact area) then I think it's really better to not check for it downstream again. It's likely harmless, but it adds the illusion that there is some possibility that equals might fail, and makes it harder to understand how the code works. 

In short, it just gets confusing. You've got a comment here, which satisfys any concerns, but just wanted to bring that up incase it comes up again.

File: fda-model/metrics/ConsequenceResults.cs (hunk 9 above)
Comment (Brennan1994) [reply fda-model/metrics/ConsequenceResults.cs:39]: Why was it done in the first place? It's a really strange move that probably should have been commented, but we didn't. I'm afraid to remove it now because I don't understand why it was there.

File: fda-model/stageDamage/ImpactAreaStageDamage.cs (hunk 4 above)
Comment (Brennan1994) [reply fda-model/stageDamage/ImpactAreaStageDamage.cs:40]: totally fine with me.

File: fda-model/stageDamage/ImpactAreaStageDamage.cs [hunk 13]
```diff
@@ -58,7 +58,7 @@ public class ImpactAreaStageDamage : Validation, IReportMessage
             _DischargeStage = dischargeStage;
             _ImpactAreaID = impactAreaID;
             _inventory = inventory.GetInventoryTrimmmedToPolygon(impactAreaID);
-            _inventory = inventory;
+            //_inventory = inventory;
```
Comment (Brennan1994) [inline fda-model/stageDamage/ImpactAreaStageDamage.cs:61]: That can just be deleted.

File: fda-model/stageDamage/ImpactAreaStageDamage.cs (hunk 3 above)
Comment (Brennan1994) [reply fda-model/stageDamage/ImpactAreaStageDamage.cs:285]: If you're making something public or static for testing, make sure you leave a comment on it. We really need to be careful doing that. It makes our code confusing.  There is a way to test private methods. I've never done it, but I'll share some info in rocketchat or github.

File: fda-model/stageDamage/ImpactAreaStageDamage.cs (hunk 2 above)
Comment (Brennan1994) [reply fda-model/stageDamage/ImpactAreaStageDamage.cs:291]: refactor to be a regular for loop for(int i = 0; i < extropolated stages...... and so on. The combo for loop and counter isn't very clean, and the fix is super easy.

File: fda-model/stageDamage/ImpactAreaStageDamage.cs (hunk 5 above)
Comment (Brennan1994) [reply fda-model/stageDamage/ImpactAreaStageDamage.cs:296]: make sure this is commented, so we know we're using access modifiers and static specifically for testing. Because someone eventually (probably future us) might wonder why the hell we did that if we don't.

File: fda-model/structures/Inventory.cs [hunk 14]
```diff
@@ -15,6 +15,7 @@ public class Inventory
     private PolygonFeatureLayer _impactAreaSet;
     private List<OccupancyType> _Occtypes;
     private string _impactAreaUniqueColumnHeader;
+    private bool _FilterStructuresByImpactArea { get; } = true;
```
Comment (Brennan1994) [inline fda-model/structures/Inventory.cs:18]: Confused by this. Why is this here?

Comment (Brennan1994) [reply fda-model/structures/Inventory.cs:18]: nevermind. I get it.

**Review by Brennan1994 — CHANGES_REQUESTED** [review]

File: fda-model/metrics/ConsequenceDistributionResults.cs (hunk 8 above)
Comment (rnugent3) [reply fda-model/metrics/ConsequenceDistributionResults.cs:150]: The fix will require a complete reorganization of how we're collecting damage so that we can be agnostic to units of measurement. This has been haunting me for the last two weeks, and I have not yet figured it out 100% but after a lot of trial and error, I think I see a path. Because it is a relatively complex refactor, I think I should do this work on a fresh branch.

File: fda-model/stageDamage/ImpactAreaStageDamage.cs (hunk 2 above)
Comment (rnugent3) [reply fda-model/stageDamage/ImpactAreaStageDamage.cs:291]: done

File: fda-model/metrics/ConsequenceDistributionResults.cs (hunk 12 above)
Comment (rnugent3) [reply fda-model/metrics/ConsequenceDistributionResults.cs:317]: I will refactor.

File: fda-model/stageDamage/ImpactAreaStageDamage.cs (hunk 13 above)
Comment (rnugent3) [reply fda-model/stageDamage/ImpactAreaStageDamage.cs:61]: done

File: fda-model/structures/Inventory.cs (hunk 14 above)
Comment (rnugent3) [reply fda-model/structures/Inventory.cs:26]: deleted now!

**Comment by rnugent3** [comment]
Immediately requested changes are complete and new issues have been set up for near-term work.

**Review by Brennan1994 — APPROVED** [review]
